content/src/content/jcr_root/apps/core/wcm/components/commons/datalayer/clientlibs/js/datalayer.js

```diff
@@ -0,0 +1,170 @@
+/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ ~ Copyright 2019 Adobe
+ ~
+ ~ Licensed under the Apache License, Version 2.0 (the "License");
+ ~ you may not use this file except in compliance with the License.
+ ~ You may obtain a copy of the License at
+ ~
+ ~     http://www.apache.org/licenses/LICENSE-2.0
+ ~
+ ~ Unless required by applicable law or agreed to in writing, software
+ ~ distributed under the License is distributed on an "AS IS" BASIS,
+ ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ ~ See the License for the specific language governing permissions and
+ ~ limitations under the License.
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+(function() {
+    "use strict";
+
+    /* eslint no-console: "off" */
+
+    // Initializes the data layer
+    function init() {
+        window.dataLayer = window.dataLayer || [];
+        populateDataLayer();
+        window.dataLayer.state = getComputeState(window.dataLayer);
+        overridePush();
+        populateDataLayerAfterOverride();
+        console.log("data layer script initialized");
+    }
+
+    // Returns the computed state by iterating over all the events stored in the data layer array
+    function getComputeState(dataLayer) {
+        var state = {};
+        dataLayer.forEach(function(event) {
+            updateState(state, event);
+        });
+        return state;
+    }
+
+
+    // Augments the push function (only for the data layer object) to also update the data layer state
+    function overridePush() {
+        window.dataLayer.push = function() {
+            var event = arguments[0];
+            updateState(window.dataLayer.state, event);
```

Comment:
we should apply this to all arguments of the push so you can do something like:
window.dataLayer.push({
    "type": "my event 1"
    "data": {}
}, {
    "type": "my event 2"
    "data": {}
});

Reply:
And we could wrap the handling in a single function that can do all the necessary tasks (which will likely be more than simply updating the state, such as event handling, validating the event etc.) before calling the prototype push that adds it to the array.

Reply:
Makes sense! I implemented it in a recent commit.